baremetal/metal3machine_manager.go

```diff
@@ -1262,122 +1281,208 @@ func (m *MachineManager) GetProviderIDAndBMHID() (string, *string) {
-		}
+	m.SetProviderID(node.Spec.ProviderID)
+	m.SetReadyTrue()
+	return nil
+}
+
+func (m *MachineManager) NodeWithMatchingProviderIDExists(ctx context.Context, clientFactory ClientGetter) bool {
+	if !m.Metal3MachineHasProviderID() {
+		return false
 	}
 
-	if matchingNodesCount == 1 {
-		return nil
+	providerIDLegacy, providerIDNew, err := m.getPossibleProviderIDs(ctx)
+	if err != nil {
+		return false
 	}
-	nodes, countNodesWithLabel, err := m.getNodesWithLabel(ctx, nodeLabel, clientFactory)
+
+	node, err := m.getNodeByProviderID(ctx, providerIDLegacy, providerIDNew, clientFactory)
 	if err != nil {
 		errMessage := "error retrieving node, requeuing"
 		m.Log.Info(errMessage)
+		return false
+	}
 
-		return WithTransientError(errors.New(errMessage), requeueAfter)
+	m.Log.Info("matching node found", "node", node.GetName(), "providerID", node.Spec.ProviderID)
+	return true
+}
+
+// SetProviderIDFromNodeLabel finds a Node by label and sets ProviderID on it.
+func (m *MachineManager) SetProviderIDFromNodeLabel(ctx context.Context, clientFactory ClientGetter) (success bool, err error) {
+	corev1Remote, err := clientFactory(ctx, m.client, m.Cluster)
+	if err != nil {
+		return false, errors.Wrap(err, "Error creating a remote client")
 	}
-	if countNodesWithLabel == 0 {
+	bmhUID, err := m.getBmhUIDFromM3Machine(ctx)
+	if err != nil {
+		errMessage := "unable to retrieve BMH UID from Metal3Machine"
+		m.Log.Info(errMessage)
+		return false, WithTransientError(errors.New(errMessage), requeueAfter)
+	}
+
+	nodeLabel := fmt.Sprintf("%s=%s", ProviderLabelPrefix, bmhUID)
+	nodes, countNodesWithLabel, err := m.getNodesWithLabel(ctx, nodeLabel, clientFactory)
+	if err != nil {
+		errMessage := fmt.Sprintf("error retrieving node with label %s, requeuing", nodeLabel)
+		m.Log.Info(errMessage)
+		return false, WithTransientError(errors.Wrap(err, errMessage), requeueAfter)
+	}
+	if countNodesWithLabel == 0 && m.Machine.Spec.Bootstrap.ConfigRef != nil {
 		// The node could either be still running cloud-init or have been
 		// deleted manually. TODO: handle a manual deletion case.
 		errMessage := "requeuing, could not find node with label: " + nodeLabel
 		m.Log.Info(errMessage)
-		return WithTransientError(errors.New(errMessage), requeueAfter)
+		return false, WithTransientError(errors.New(errMessage), requeueAfter)
 	}
 	if countNodesWithLabel > 1 {
-		return errors.Wrap(err, fmt.Sprintf("Found multiple target nodes with the same label: (%s)", nodeLabel))
+		return false, errors.Wrap(err, fmt.Sprintf("Found multiple target nodes with the same label: (%s)", nodeLabel))
 	}
-	var nodeVar corev1.Node
-	for _, node := range nodes.Items {
-		oldData, err := json.Marshal(node)
+
+	providerIDLegacy, providerIDNew, err := m.getPossibleProviderIDs(ctx)
+	if err != nil {
+		errMessage := fmt.Sprintf("unable to retrieve BMH name from Metal3Machine: %v", err)
+		m.Log.Info(errMessage)
+		return false, WithTransientError(errors.New(errMessage), requeueAfter)
```

Comment:
nit: build the error using fmt.Errorf and %w, so that the wrapped error is stored. You can use the String() of it for logging